mythril/laser/ethereum/state.py

```diff
@@ -1,16 +1,82 @@
+from z3 import BitVec, BitVecVal, BitVecRef, BitVecNumRef, BitVecSort, Solver, ExprRef, Concat, sat, simplify, Array, ForAll, Solver, UGT, Implies
+from z3.z3types import Z3Exception
 from mythril.disassembler.disassembly import Disassembly
 from copy import copy, deepcopy
 from enum import Enum
 from random import randint
+from mythril.laser.ethereum.util import get_concrete_int
 
 from mythril.laser.ethereum.evm_exceptions import StackOverflowException, StackUnderflowException
 
-
 class CalldataType(Enum):
     CONCRETE = 1
     SYMBOLIC = 2
 
+class Calldata:
+
+    def __init__(self, tx_id, starting_calldata: bytes=None):
+        self.tx_id = tx_id
+        if starting_calldata:
+            self._calldata = []
+            self.calldatasize = BitVecVal(len(starting_calldata), 256)
+            self.concrete = True
+        else:
+            self._calldata = Array('{}_calldata'.format(self.tx_id), BitVecSort(256), BitVecSort(8))
+            self.calldatasize = BitVec('{}_calldatasize'.format(self.tx_id), 256)
+            self.concrete = False
+
+        self.starting_calldata = starting_calldata if starting_calldata else []
+
+    def set_global_state(self, state):
+        self.state = state
+        constraints = []
+
+        if self.concrete:
+            for cd_byte in self.starting_calldata:
+                if type(cd_byte) == int:
+                    self._calldata.append(BitVecVal(cd_byte, 8))
+                else:
+                    self._calldata.append(cd_byte)
+            constraints.append(self.calldatasize == len(self.starting_calldata))
+        else:
+            x = BitVec('x', 256)
+            constraints.append(ForAll(x, Implies(self[x] != 0, UGT(self.calldatasize, x))))
+
+        self.state.mstate.constraints.extend(constraints)
+
+    def concretized(self, model):
+        result = []
+        for i in range(get_concrete_int(model.eval(self.calldatasize, model_completion=True))):
+            result.append(get_concrete_int(model.eval(self[i], model_completion=True)))
+
+        return result
+
+    def get_word_at(self, index: int):
+        return self[index:index+32]
+
+    def __getitem__(self, item):
+        if isinstance(item, slice):
+            stop = item.stop
+            if self.concrete and type(stop) == int:
+                calldatasize = get_concrete_int(self.calldatasize)
+                if stop > calldatasize:
+                    stop = calldatasize
+
+            try:
+                current_index = item.start if type(item.start) in [BitVecRef, BitVecNumRef] else BitVecVal(item.start, 256)
+                dataparts = []
+                while simplify(current_index != stop):
+                    dataparts.append(self[current_index])
+                    current_index = simplify(current_index + 1)
+            except Z3Exception:
+                raise IndexError("Invalid Calldata Slice")
+
+            return simplify(Concat(dataparts))
+
+        if self.concrete:
```

Comment:
Given concrete calldata initialized with [1, 3, 5]
What happens if I do:
calldata[10] ?
It seems like this will return an index error, while it should return 0

Reply:
Fixed.